python/pyspark/sql/metrics.py

```diff
@@ -0,0 +1,238 @@
+        Converts the collected metrics into a dot representation. Since the graphviz Digraph
+        implementation provides the ability to render the result graph directory in a
+        notebook, we return the graph object directly.
+
+        If the graphviz package is not available, a PACKAGE_NOT_INSTALLED error is raised.
+
+        Parameters
+        ----------
+        filename - str, optional
+            The filename to save the graph to given an output format. The path can be
+            relative or absolute.
+
+        out_format - str
+            The output format of the graph. The default is 'png'.
+
+        Returns
+        -------
+
+        """
+        try:
+            import graphviz
+
+            dot = graphviz.Digraph(
+                comment="Query Plan",
+                node_attr={
+                    "shape": "box",
+                    "font-size": "10pt",
+                },
+            )
+
+            root, graph = self.extract_graph()
+            for k, v in graph.items():
+                # Build table rows for the metrics
+                rows = "\n".join(
+                    [
+                        (
+                            f'<TR><TD><FONT POINT-SIZE="8">{x.name}</FONT></TD><TD>'
+                            f'<FONT POINT-SIZE="8">{x.value} ({x.metric_type})</FONT></TD></TR>'
+                        )
+                        for x in v.metrics
+                    ]
+                )
+
+                dot.node(
+                    str(k),
+                    """<<TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0">
+                    <TR>
+                        <TD COLSPAN="2" BGCOLOR="lightgrey">
+                            <FONT POINT-SIZE=\"10\">{}</FONT>
+                        </TD>
+                    </TR>
+                    <TR><TD COLSPAN="2"><FONT POINT-SIZE=\"10\">Metrics</FONT></TD></TR>
+                    {}
+                    </TABLE>>""".format(
+                        v.name, rows
+                    ),
+                )
+                for c in v.children:
+                    dot.edge(str(k), str(c))
+
+            if filename:
+                dot.render(filename, format=out_format, cleanup=True)
+            return dot
+
+        except ImportError:
+            raise PySparkValueError(
+                error_class="PACKAGE_NOT_INSTALLED",
+                message_parameters={"package_name": "graphviz", "minimum_version": "0.20"},
+            )
+
+
+class QueryExecution:
+    """The query execution class allows users to inspect the query execution of this particular
+    data frame. This value is only set in the data frame if it was executed."""
+
+    def __init__(
+        self, metrics: Optional[list[PlanMetrics]], obs: Optional[Sequence[ObservedMetrics]]
+    ):
+        self._metrics = CollectedMetrics(metrics) if metrics else None
+        self._observations = obs if obs else []
+
+    @property
+    def metrics(self) -> Optional[CollectedMetrics]:
```

Comment:
I think we should probably match the naming w/ Scala side `QueryExecution` ..

Reply:
In Scala, I can use the `queryExecution` field to get analyzed, optimized, and executed plans. These are very useful for debugging. Are we planning to add them to this class, too?

Reply:
The metrics that we observed here are generated from the query result without any additional RPC. I don't think that accessing the analyzed, optimized, executed plans directly (outside of explain) is really a good CUJ here.

Reply:
Does Scala client have `df.queryExecution`? I don't see it in the Scala client.
If not, I guess we should avoid the same name as it's confusing because it's missing many properties that exist in the classic's `df.queryExecution`.

I'd rather prefer `CollectedMetrics` and it can be added to the classic as well to be consistent.